### DISCUSSION TITLE
Building a cFS executable for VxWorks 7.0 on ARM v8

### DISCUSSION THREAD

**Discussion by asvatos1** [nasa/cFS#356] [Q&A] [answered]
Hi,

I am working in Ubuntu 20.04 where I am cross-compiling a cFS executable to run on VxWorks 7.0 running on an ARM v8 (Cortex-A72 processor) and I would like to know if there is a working example available somewhere for how to build cFS to run on VxWorks.

From reviewing [another post](https://github.com/nasa/cFS/discussions/314), I have gathered that the steps to build cFS to run on VxWorks are:

1. Modify targets.cmake to include the line 'SET(cpu1_SYSTEM arm-vxworks7.0)' to force the build system to search for that toolchain file.
2. Create the requisite toolchain file with the name 'arm-vxworks7.0'.
3. No need to modify OSAL BSP for generic-vxworks as it generic enough to be used on multiple platforms.
4. Possibly modify the PSP since it is for Power PC.
3. Call 'make prep arm-vxworks7.0' from console in the cFS directory.

To accomplish (2), I examined the toolchain-ppc-vxworks6.9.cmake toolchain file and I am modifying it by hand be suitable for my case. 

Is this the correct approach? 

Thank you for reviewing this post.

-Adam

**Comment by asvatos1** [comment]
My assumption about there being no need to modify the OSAL or the OSAL board support package was incorrect. From the VxWorks site:
```
VLegacy BSPs (VxWorks 6.9.x.x and older) are not compatible with VxWorks 7 due to the following fundamental differences:
	VxWorks 7 uses a new VxBus driver infrastructure that, among other improvements, is configured based on device probing rather than hard coding of device information.
	VxWorks 7 BSPs use a completely different directory structure.
	VxWorks 7 BSPs provide separation between processor family-specific components and board-specific components.
```
While building the cFS executable according to my requirements, I found that many of the VxWorks definitions such as SEM_ID, VX_BINARY_SEMAPHORE couldn't be found. This is because the names of these definitions have changed from VxWorks 6.9 to 7.0. To solve that one would need to re-write OSAL to correctly interface with VxWorks 7.0.

**Comment by skliper** [comment, marked as answer]
There was some work done in the "Caelum" development cycle to get us closer to supporting VxWorks 7 (https://github.com/nasa/osal/issues/610 for example), and I've heard of some prototyping based on the SDK.  VxWorks 7 support out of the box is definitely on the roadmap, and we are actively soliciting contributions and pursuing collaboration opportunities to get this done.  Eventually we'd like to get a VxWorks 7 SDK workflow into our CI.  I realize this isn't immediate help, but perhaps there's enough interest across the community that we can add to the current momentum around this topic!

**Reply by pmehradnia** [reply]
Hi @skliper could you let me know if there has been any updates regarding VxWorks 7 support and potential future releases supporting it?

**Reply by skliper** [reply]
In theory there should be nothing in cFE or OSAL preventing compiling/running with VxWorks 7.  Note there is an open PR in the PSP (https://github.com/nasa/PSP/pull/193) that needs rework to actually get accepted to support VxWorks 7.  @astrogeco may be able to provide an update on the prototyping effort with the SDK (I haven't seen any related pull requests).  I'm not aware of dedicated resources available to the core team to support the upgrade and add into CI, so we are currently reliant on external contributions.  Happy to help coordinate/assist if there's sufficient external interest!

**Reply by pmehradnia** [reply]
@skliper Thanks for the update and the quick reply! 👍

**Reply by edison369** [reply]
Hi @skliper! Do you know if it is already posible to build a cFS executable using the VxWorks SDK? If not, I must have the licensed VxWorks installation to do it, right?

**Reply by skliper** [reply]
@edison369 I've **heard** someone here has built and run cFS using the VxWorks SDK here, but I haven't gotten more details than that.  I'll reach out and see if I can find out more details.

**Reply by astrogeco** [reply]
Hi from the other side! @ezpollack and @acudmore looked into this a while ago and made some progress using the ROS2 VxWorks repo https://github.com/Wind-River/vxworks7-ros2-build as a starting point. I vaguely remember that the "free" SDK was missing some features needed to for OSAL but that's as far as my memory goes. Good luck!

**Reply by edison369** [reply]
@skliper and @astrogeco thank you both for the quick reply and your help! I'll try then 😃
